### DISCUSSION TITLE
Is it possible to validate objects that lack a required property with a default value?

### DISCUSSION THREAD

**Discussion by dhalfageme** [json-schema-org/community#139] [Q&A] [answered]
Hi!. Lets say I have this schema:

```{
  "$schema": "http://json-schema.org/2020-12/schema#",
  "$id": "test",
  "title": "Product",
  "description": "A product from the catalog",
  "type": "object",

  "properties": {
    "complexObject": {
      "type": "object",
      "required": ["id"],
      "properties": {
        "id": {
          "description": "The unique identifier for a product",
          "type": "integer",
          "default": 2
        },
        "name": {
          "description": "Name of the product",
          "type": "string"
        }
      }
    }
  }
}
```

And this object.

```
{
  "complexObject": { "name":  "product name"}
}
```

The object doesn't validate using everit-org/json-schema  java validor because it lacks of `complexObject.id`property, which has a default value. I put a nested property to be as much general as possible.

I would like to pass validation and build a new JSON object using default values for missing properties.

Thanks.

**Comment by gregsdennis** [comment, marked as answer]
The `default` keyword doesn't perform any validation. It's an annotation-only keyword. This means that its purpose is to provide it's value to the application, and the application should determine what to do with that information.

Thus, your instance doesn't validate because `id` is absent.

My suggestion would be to make `id` optional (just don't list it in `required`), which would allow your instance to pass.  One the validation has passed, it should be accompanied by any annotations.  From there, you can decide what to do with them.

**Comment by dhalfageme** [comment]
@gregsdennis Sorry for the out of scope response to your original answer

So if I would like to do this default population, because it's a requirement, would I have to make some custom implementation that creates the properties when applying the new schema? There's no a workaround for this, right?

 I have hundreds of objects that I would like to evolve to new schema definitions and provide valid values when reading them

**Reply by gregsdennis** [reply]
You might be able to find an implementation that has an option to do this for you, but you'll have to search.

The best approach is to not rely on the validator to enforce defaults and handle the default in code.

**Reply by karenetheridge** [reply]
You can use the annotations for that (and that's what they're there for! :)) -- after evaluating, you can take the list of annotations that is returned by the evaluator and update your data instance with those default values.

**Reply by jdesrosiers** [reply]
I don't think you can use annotation results from the output format for populating `default`s. You don't have that information until after you've validated, but you would need to fill in defaults before validation in this case.

**Reply by karenetheridge** [reply]
Yes, I mean that you evaluate the schema, then collect the annotations and alter the data, and then use the new data with defaults in your application to do whatever you need to do.

**Reply by gregsdennis** [reply]
It's important that the schema can't _require_ a field that has a default.  Otherwise the validation will fail.

**Reply by dhalfageme** [reply]
Sorry @gregsdennis but I don't understand you. Required and default are perfectly compatible in the schema. At least I've tried to validate and object with a required property with a default, with a different value, and it passes the validation.

@karenetheridge The annotations usage you propose required some kind of POJO that needs be defined prior to the validation and so on?

**Reply by gregsdennis** [reply]
What I mean is that by declaring a `default` and expecting its value to be applied, you _expect_ that the property may not be in the JSON instance.  But `required` implies that the property _is_ in the JSON instance.  You can't have it both ways.  Either the JSON instance has the property (satisfying `required`) or it doesn't (meaning `default` should be applied).

The `required` keyword knows nothing of `default` or any other keyword.  It simply validates that the JSON instance contains those properties.

If you want defaults to be applied _and_ `required` to validate them, then you need to pre-process the defaults into the JSON instance before performing validation.

**Reply by Relequestual** [reply]
> Yes, I mean that you evaluate the schema, then collect the annotations and alter the data, and then use the new data with defaults in your application to do whatever you need to do.

Is this something your implementation can do? Collect annotations by evaluation without an instance?

I guess it wouldn't actually be that complex to do unless references are used.

**Reply by dhalfageme** [reply]
Yeah @Relequestual, I am looking for some way to perform that... If  you @karenetheridge know a way to get this working, it is now the real problem, not only to make the validation pass, also to provide valid values when reading the objects that lack properties with defaults.

**Reply by Relequestual** [reply]
[AJV can do this out of the box](https://ajv.js.org/guide/modifying-data.html#assigning-defaults), but I don't know what language you're working with.

**Reply by dhalfageme** [reply]
I am using Java

**Reply by awwright** [reply]
For some more discussion on why "default" doesn't work this way (by default), see https://github.com/json-schema-org/json-schema-spec/issues/867
